protocol/directory.go

```diff
@@ -0,0 +1,190 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ConiksDirectory struct {
+	pad      *merkletree.PAD
+	useTBs   bool
+	tbs      map[string]*merkletree.TemporaryBinding
+	policies merkletree.Policies
+}
+
+func InitDirectory(epDeadline merkletree.TimeStamp, vrfKey vrf.PrivateKey,
+	signKey sign.PrivateKey, dirSize uint64, useTBs bool) *ConiksDirectory {
+	d := new(ConiksDirectory)
+	d.SetPolicies(epDeadline, vrfKey)
+	pad, err := merkletree.NewPAD(d.policies, signKey, dirSize)
+	if err != nil {
+		panic(err)
+	}
+	d.pad = pad
+	d.useTBs = useTBs
+	if useTBs {
+		d.tbs = make(map[string]*merkletree.TemporaryBinding)
+	}
+	return d
+}
+
+func (d *ConiksDirectory) Update() {
+	d.pad.Update(d.policies)
+	// clear issued temporary bindings
+	for key := range d.tbs {
+		delete(d.tbs, key)
+	}
+}
+
+func (d *ConiksDirectory) SetPolicies(epDeadline merkletree.TimeStamp, vrfKey vrf.PrivateKey) {
+	d.policies = merkletree.NewPolicies(epDeadline, vrfKey)
+}
+
+func (d *ConiksDirectory) LatestSTR() *merkletree.SignedTreeRoot {
+	return d.pad.LatestSTR()
+}
+
+// HandleOps validates the request message and then pass it to
+// appropriate operation handler according to the request type.
+func (d *ConiksDirectory) HandleOps(req *Request) (Response, ErrorCode) {
```

Comment:
The main reason for this is to prevent invalid requests from being passed into the handlers. Since the `server` package is just a reference implementation, so I think it'd be better if the library provided an API for validation.